File: internal/model/game_challenge.go
package model

import "gorm.io/gorm"

type GameChallenge struct {
	ID          uint       `json:"id,omitempty"`
	GameID      uint       `gorm:"uniqueIndex:game_challenge_idx" json:"game_id,omitempty"`
	Game        *Game      `json:"game,omitempty"`
	ChallengeID uint       `gorm:"uniqueIndex:game_challenge_idx" json:"challenge_id,omitempty"`
	Challenge   *Challenge `json:"challenge,omitempty"`
	IsEnabled   *bool      `gorm:"default:false;not null;" json:"is_enabled,omitempty"`
	Pts         int64      `gorm:"-" json:"pts,omitempty"`
	MaxPts      int64      `gorm:"default:1000;not null;" json:"max_pts,omitempty"`
	MinPts      int64      `gorm:"default:200;not null;" json:"min_pts,omitempty"`
}

func (g *GameChallenge) BeforeDelete(db *gorm.DB) (err error) {
	db.Table("submissions").Where("game_id = ?", g.GameID).Where("challenge_id = ?", g.ChallengeID).Delete(&Submission{})
	db.Table("notices").Where("game_id = ?", g.GameID).Where("challenge_id = ?", g.ChallengeID).Delete(&Notice{})
	return nil
}

File: web/src/types/game_challenge.ts
import { Challenge } from "./challenge";
import { Game } from "./game";

export interface GameChallenge {
    challenge_id?: number;
    challenge?: Challenge;
    game_id?: number;
    game?: Game;
    is_enabled?: boolean;
    difficulty?: number;
    max_pts?: number;
    min_pts?: number;
    first_blood_reward_ratio?: number;
    second_blood_reward_ratio?: number;
    third_blood_reward_ratio?: number;
}

export interface GameChallengeUpdateRequest {
    challenge_id?: number;
    game_id?: number;
    is_enabled?: boolean;
    max_pts?: number;
    min_pts?: number;
    difficulty?: number;
    first_blood_reward_ratio?: number;
    second_blood_reward_ratio?: number;
    third_blood_reward_ratio?: number;
}

export interface GameChallengeDeleteRequest {
    challenge_id?: number;
    game_id?: number;
}

export interface GameChallengeCreateRequest {
    game_id?: number;
    challenge_id?: number;
    is_enabled?: boolean;
    max_pts?: number;
    min_pts?: number;
    difficulty?: number;
    first_blood_reward_ratio?: number;
    second_blood_reward_ratio?: number;
    third_blood_reward_ratio?: number;
}
